### PR TITLE
KAFKA-1893: Allow regex subscriptions in the new consumer

### PR DESCRIPTION
**Pull request by SinghAsDev**

### CODE REVIEW

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 1]
```diff
@@ -660,6 +670,85 @@ public void subscribe(TopicPartition... partitions) {
...
+        } finally {
+            release();
+        }
+    }
+
+    private void schedulePatternSubscriptionTask(final long interval) {
+        this.patternSubscriptionTask = new DelayedTask() {
+            public void run(long now) {
+                List<String> topicsToSubscribe = new ArrayList<>();
+                for (String topic: listTopics().keySet()) {
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: It would be nice to keep this asynchronous. Instead of calling listTopics, which blocks, could we initiate a new metadata fetch directly and add a listener to handle the completion?

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Hey @hachikuji , even I was initially thinking to have it asynchronous, but I could not convince myself that we will gain much with it? schedulePatternSubscriptionTask is already a delayed task that user do not have to wait on. Moreover initiating a new metadata fetch directly needs to handle retries, so more params for max retries, interval, etc. I was under impression that we made ListTopics handle all that so that we will not have to worry about it again during regex implementation. I might be missing something.

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @SinghAsDev Since KafkaConsumer has only one thread, even scheduled tasks have to be executed in that thread, which means the user has to wait for them. Since you can't really control when the tasks will be executed, in the worst case, it could turn a non-blocking call into a blocking one. And I don't see why error handling can't be handled asynchronously. For updating regex subscriptions, I wouldn't think it too much of a big deal even if we just ignored failures and waited for the next metadata update, though it would be easy to implement retries with backoff (I think we do this for heartbeats already).

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: True. For error handling, I was saying that it might require more configs not the complexity. However, we should be able to reuse the existing configs. Will update the patch. Thanks!

**Comment by benstopford** [comment]
Good comments from Grant. Other than that this change looks good.

**Comment by SinghAsDev** [comment]
@hachikuji @granthenke I have addressed your review comments. Let me know if I am still missing something. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 2]
```diff
@@ -1171,4 +1253,43 @@ private void release() {
...
+
+        public void run(final long now) {
+            final List<String> topicsToSubscribe = new ArrayList<>();
+
+            final RequestFuture<ClientResponse> requestFuture = fetcher.sendMetadataRequest();
+            if (requestFuture != null) {
+                requestFuture.addListener(new RequestFutureListener<ClientResponse>() {
+                    @Override
+                    public void onSuccess(ClientResponse value) {
+                        MetadataResponse response =
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @SinghAsDev Are there any error cases to check here? Disconnects for example?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 3]
```diff
@@ -1171,4 +1253,43 @@ private void release() {
...
+                            new MetadataResponse(requestFuture.value().responseBody());
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: I think requestFuture.value() is the same as value.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 2 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Do you mean in case of errors, I should log it and schedule the task as in case of onFailure?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 3 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: True. Will make the change. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 2 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @SinghAsDev Haha, I can understand why this is a little confusing, but the network layer considers the request a "success" if it gets a response. However, that doesn't mean that there wasn't an error code in that response. It might be nice to handle this at the network layer, but it seems to me that there wasn't a generic way to check for errors. Each response object had the error code at a different location in its schema, so the only thing we could do is pass the response back and let the application determine if there was an error. The one case where we might be able to infer errors generically is by checking ClientResponse.wasDisconnected(), but I don't think we do this either (I've forgotten if there's a good reason for that).

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @hachikuji updated. Let me know if I got that right :)

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @SinghAsDev I think that works! I wasn't sure if we needed to handle errors in the MetadataResponse itself, but it looks like the topics are only added to the Cluster if they had no error in the response.

**Comment by hachikuji** [comment]
LGTM

@guozhangwang Do you want to have a look?

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java [hunk 4]
```diff
@@ -165,6 +165,10 @@
     /** <code>connections.max.idle.ms</code> */
     public static final String CONNECTIONS_MAX_IDLE_MS_CONFIG = CommonClientConfigs.CONNECTIONS_MAX_IDLE_MS_CONFIG;
 
+    /** <code>pattern.subscription.interval.ms</> **/
+    public static final String PATTERN_SUBSCRIPTION_INTERVAL_MS_CONFIG = "pattern.subscription.interval.ms";
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Do we need a new config for subscription interval? Could we piggy-back the logic on regular metadata refresh?

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 5]
```diff
@@ -660,6 +674,74 @@ public void subscribe(TopicPartition... partitions) {
...
+     * <li>Number of partitions change for any of the subscribed list of topics
+     * <li>Topic is created or deleted
+     * <li>An existing member of the consumer group dies
+     * <li>A new member is added to an existing consumer group via the join API
+     * </ul>
+     *
+     * @param pattern Pattern to subscribe to
+     */
+    @Override
+    public void subscribe(Pattern pattern) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Right now the subscribe(Pattern) is overriden while subscribe(Topic / Partition) is incremental. Could we allow users to do:

subscribe("PageView_");
subscribe("ProfileView_");

incrementally to be consistent with others?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @guozhangwang In KAFKA-2388, I think the plan is to remove the ability to subscribe incrementally (instead you have to provide the full list), so this would be consistent if we end up accepting that proposal.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 6]
```diff
@@ -660,6 +674,74 @@ public void subscribe(TopicPartition... partitions) {
...
+    private boolean isAlreadySubscribed(String topic) {
+        return this.subscriptions.subscribedTopics().contains(topic);
+    }
+
+    /**
+     * Unsubscribe from topics subscribed via pattern subscription
+     *
+     * @param pattern Pattern to unsubscribe from, must match the pattern subscribed to
+     */
+    public void unsubscribe(Pattern pattern) {
```
Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: I originally thought the semantics of unsubscribe() is blacklisting. For example:

subscribe("PageView*")
unsubscribe("PageViewTopicA")

will subscribe to all page-view topics except page-view-A. Arguably this can be done solely in subscribe() with"^" in regex, but I feel this semantics is more natural for users.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Hi @guozhangwang , I had some thoughts along the same line and have posted them on the JIRA. Below is the copy.

"Should we support topic and regex subscription at same time? There are a few special cases that will need attention if we decide to have regex and topic subscription work at same time. Say one subscribed to test\* and then later explicitly unsubscribed from test1. test1 will get subscribed the next time regex check is run. This can be handled by maintaining a blacklist for a particular pattern. The point is we can support that, however it will have complicated logic. Wondering what will we gain from this complicated logic though. One can have a pattern to subscribe to individual topic or a collection of topics or both. Having support for topic and regex subscription does not buy a lot. If nobody suggests otherwise I will go with making topic and regex subscription mutually exclusive.
Also, in my current implementation I am only supporting subscribing to a single pattern and not multiple patterns as multiple patterns can be combined into a single pattern. Hope that is fine as well."

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 5 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Also, with regex one can simply subscribe to "(PageView)|(ProfileView)" instead of incrementally subscribing to a list of regex. I was not aware of what @hachikuji mentioned, but if that is the plan then it takes away your concern of consistency. Let us know your thoughts @guozhangwang

File: clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java (hunk 4 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/ConsumerConfig.java]: Makes sense. Removed.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 6 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @SinghAsDev @guozhangwang I wonder if we could support blacklist filtering through a second argument to subscribe. For example:

``` java
void subscribe(Pattern whitelist, Pattern blacklist);
```

This might get around some of the complexity mentioned above.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: This will still require incremental support, right?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Would it have to? I would see each call to subscribe as replacing the previous one. Perhaps if we wanted to support multiple regex subscriptions, we could do:

``` java
void subscribe(List<Pattern> patterns);
```

And we could require users to use '^' for blacklisting. This would be consistent with the changes proposed in KAFKA-2388.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: I like "each call to subscribe as replacing the previous one" better than `void subscribe(List<Pattern> patterns)`. As later can run into issues where a pattern added later can negate the effect of pattern added before. In general, I think list of patterns is a bad idea. Would you agree?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Yeah, that's a good point. Scratch that idea.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: @guozhangwang if you think what @hachikuji suggested to support blacklist makes sense, I will go ahead and make required changes. It will be nice if you can express your thoughts on open discussions on this PR, so that I can address all issues in a single commit.

**Comment by hachikuji** [comment]
@SinghAsDev Another problem that @guozhangwang and I briefly talked about was whether we should augment the Metadata object to support getting metadata for all topics (with a hook to get a notification when there is a change)*. Your approach is exactly the one I had in mind, but it does result in a little more overhead since metadata refreshes are done both in KafkaConsumer and in NetworkClient. The nice thing about separating them is that it lets you refresh the full metadata list less frequently. If your regex only covers a small set of the topics, then that makes "normal" metadata refreshes cheaper and they can be done more often. However, it seems a little questionable whether this capability is actually useful in practice, so reducing the metadata overhead might be the bigger win. What do you think? 

*Note that it's still useful to be able to bypass NetworkClient to support the listTopics API (which doesn't affect subscriptions).

File: clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java [hunk 7]
```diff
@@ -55,17 +56,35 @@ public MockConsumer(OffsetResetStrategy offsetResetStrategy) {
...
+    @Override
+    public void subscribe(Pattern pattern) {
+        ensureNotClosed();
+        this.subscriptions.subscribe(pattern);
+        String[] topicsToSubscribe = new String[partitions.size()];
+        int pos = 0;
+        for (String topic: partitions.keySet()) {
+            if (pattern.matcher(topic).matches() &&
+                !subscriptions.subscribedTopics().contains(topic))
+                topicsToSubscribe[pos] = topic;
```
Comment (eribeiro) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: @SinghAsDev you are not increasing `pos`! It should be `topicsToSubscribe[pos++] = topic`, right?

In fact, I would use a list as below:

```
List<String> topicsToSubscribe = new ArrayList<>(partitions.size());

blahblahblah
....

subscribeTopics(topicsToSubscribe.toArray(new String[0]), true);
```

But I am fine with the array too.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/MockConsumer.java]: Aha, nice catch. Will fix. Thanks!

**Comment by SinghAsDev** [comment]
@hachikuji that makes sense and I actually [suggested](https://issues.apache.org/jira/browse/KAFKA-2275?focusedCommentId=14634080&page=com.atlassian.jira.plugin.system.issuetabpanels:comment-tabpanel#comment-14634080) the same even for listTopics patch. Is my understanding correct that what you are suggesting is similar to my suggestion on listTopics JIRA?

**Comment by hachikuji** [comment]
@SinghAsDev I think the listTopics approach is still fine and doesn't need to be changed. The key difference is that listTopics doesn't affect subscriptions, so it should not have any impact on the Metadata shared throughout the consumer.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 6 above)
Comment (ijuma) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: This is an interesting discussion. I don't have the full context of these discussions, but I can see the argument for keeping things simple and letting users rely on the regular expression for achieving their goals. I am not sure if it's any simpler to provide two separate regular expressions than just one with the appropriate negation. The latter is easier to write a unit test for too.

**Comment by guozhangwang** [comment]
@SinghAsDev Sorry for being late on this, regarding the subscribe semantics I am convinced 1) we do not need to have an extra blacklist pattern in the API but just let users specify that in the regex, and 2) we do not need to do incremental subscription and replacing subscription should just be fine.

Regarding the metadata refresh, what I was thinking is that the current patch took a different scheduled task with different topic sets outside Metadata for regex subscription, while if we can piggy-back it with metadata refresh we can potentially reduce the code complexity while the cost of having metadata refresh always asking for all topics seems OK to me since in practice users would probably want to set the "regex refreshing interval" to be the same as "metadata refreshing interval" the same anyways. What do you think?

On the other hand, we do not necessarily need to make ListTopics API also changing the Metadata states object as @hachikuji suggested since I feel it is in many cases a one-time thing.

I could review this PR again and check it in once it gets rebased / updated.

**Comment by SinghAsDev** [comment]
@guozhangwang @hachikuji makes sense. Will update. However, for this to happen we need to somehow make networkClient aware of the fact that if pattern subscription is being used get metadata for all topics. There are a few ways I could think of for doing this.
1. Have Metadata object maintain a flag that indicates wether we need metadata for all topics or just for the ones maintained in metadata.
2. Have NetworkClient maintain a flag which is set by KafkaConsumer during subscribe(pattern) and unsubscribed during unsubscribe(pattern). This will require changing the KafkaClient interface though.

I am more inclined towards Option 1. What do you guys suggest?

**Comment by hachikuji** [comment]
@SinghAsDev Option 1 definitely sounds nicer to me.

**Comment by hachikuji** [comment]
@SinghAsDev Btw, I was planning to use a MetadataListener in KAFKA-2464 to hook into metadata updates: https://github.com/apache/kafka/pull/165/files#diff-62bba39339405475f71241a182ef9819R229. Seems like that might be sufficient for this case too.

**Comment by SinghAsDev** [comment]
@hachikuji yea, it will be needed. Should I wait for your patch to go in?

**Comment by hachikuji** [comment]
@SinghAsDev Nah, yours is probably easier to get through, so go ahead. I'll rebase accordingly.

**Comment by guozhangwang** [comment]
+1 on option 1 also.

**Comment by SinghAsDev** [comment]
@hachikuji @guozhangwang I have updated the PR with the discussed approach. Let me know how it looks. I will add some unit tests for the new methods added to metadata tonight.

**Comment by SinghAsDev** [comment]
Ahh.. looks like I will again have to rebase. Will do that along with adding unit tests for Metadata.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java [hunk 8]
```diff
@@ -183,4 +197,50 @@ public synchronized long lastSuccessfulUpdate() {
     public long refreshBackoff() {
         return refreshBackoffMs;
     }
+
+    /**
+     * Set state to indicate if metadata for all topics in Kafka cluster is required or not.
+     * @param needMetadaForAllTopics boolean indicating need for metadata of all topics in cluster.
+     */
+    public void needMetadataForAllTopics(boolean needMetadaForAllTopics) {
+        this.needMetadataForAllTopics = needMetadaForAllTopics;
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: I wonder if we should clear the topic collection when this is set. If we did that, then the change in NetworkClient would be unnecessary. Is there any reason not to?

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: This will be called only when a pattern is subscribed. At that time yes, what you are saying makes sense. However, in subsequent metadata updates the changes in NetworkClient is still needed as metadata will have topics that matched pattern and were added in last update. Makes sense?

**Comment by SinghAsDev** [comment]
@hachikuji @guozhangwang could you guys take a look. Thanks!

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 8 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: Yeah, I guess that makes sense. It kind of feels weird to be tracking a topic list in Metadata that we're not going to use, however. My feeling is that maybe we shouldn't update Metadata for topics subscribed from a regex. Instead we just update metadata once in subscribe(pattern). Does that make sense?

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: I am a bit lost here. Even if we clear topics here, topics will be added by KafkaConsumer.onMetadataUpdate and any metadata update after that will fetch metadata for only those topics if NetworkClient's changes are removed. I think I am missing something here :)

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: Probably my fault. :)  I was trying to suggest to not have KafkaConsumer.onMetadataUpdate modify the Metadata topic list, just the subscription topic list. So basically in KafkaConsumer.subscribe(pattern), we set the Metadata allTopics flag, and we don't modify Metadata again (unless the user changes their subscription). Does that make any sense?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: This might not work as well as I thought since the metadata is also updated in partitionsFor(). I actually think it would be better for partitionsFor() to use the same approach as listTopics(), but that is out of scope for this issue.

File: clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java [hunk 9]
```diff
@@ -56,6 +57,16 @@
     public void assign(List<TopicPartition> partitions);
 
     /**
+    * @see KafkaConsumer#subscribe(Pattern)
+    */
+    public void subscribe(Pattern pattern);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: To be consistent with KAFKA-2388, this should probably be:

``` java
void subscribe(Pattern pattern, ConsumerRebalanceListener listener);
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java [hunk 10]
```diff
@@ -118,6 +130,24 @@ public void assign(List<TopicPartition> partitions) {
         this.assignment.keySet().retainAll(this.userAssignment);
     }
 
+    public void subscribe(Pattern pattern) {
+        if (!this.subscription.isEmpty() || !this.userAssignment.isEmpty())
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Since subscribe(topics) is no longer additive, I wonder if we should have subscribe(pattern) just replace the previous topic subscription (if there was one) instead of raising the exception?

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 8 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: Haha.. I actually made the changes and realized the same.. It does make it look cleaner though. Also, one more issue would be is when subscribed via pattern, metadata will always store metadata for all topics and a fetch will return metadata for all topics, not just for the topics that are actually subscribed via pattern subscription.

For now, I think we can just leave it as it is. If you think we should refactor partitionsFor() then let me know, we probably should create a separate JIRA for it though.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java (hunk 10 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: For some reason I think explicitly unsubscribing topic before subscribing via pattern is better, leaves less room for user error. If you think it is still OK, I can make the change. Let me know.

File: clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java (hunk 9 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/Consumer.java]: Agreed. Will change.

File: clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java (hunk 10 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/internals/SubscriptionState.java]: Haha, the corollary would be that we wouldn't need unsubscribe(pattern) anymore (we got rid of unsubscribe(topics) already). My preference is probably to override, but I feel a little concern that it would be confusing for users who expect to be able to do both (the exception makes the restriction very clear). I think another option we talked about was just to have a single unsubscribe() with no arguments, which can be used to unsubscribe from either the current topic list or topic pattern. Maybe we can drag @gwenshap in here for another opinion since Guozhang is on vacation.

File: clients/src/main/java/org/apache/kafka/clients/Metadata.java (hunk 8 above)
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: Refactoring partitionsFor does seem worthwhile to me since then Metadata is only modified by subscription (and metadata for subscribed topics is all we care about in a steady state). Maybe we can leave this patch as it is and plan to fix it in that JIRA.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/Metadata.java:207]: Agreed. Created KAFKA-2506.

**Comment by ijuma** [comment]
I notice you rebased against trunk, but did you try to build it after that? There is now a `MetadataUpdater` interface and all metadata operations go via that interface in order to support cases in the broker where we don't need metadata updates to happen while using `NetworkClient`.

**Comment by SinghAsDev** [comment]
@ijuma looks like build failed during checkstyle, I did build jars before pushing that apparently does not execute checkstyle. Pushing the fix now.

I did not get your reference to `MetadataUpdater`, am I doing something wrong in my patch?

**Comment by ijuma** [comment]
@SinghAsDev, oh I see. Seems like a bug that we don't run checkstyle then. Would you mind filing a JIRA?

Ignore my other comment, I had a second look and your change is actually within `DefaultMetadataUpdater`, so it's fine.

**Comment by SinghAsDev** [comment]
@ijuma created KAFKA-2513.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 11]
```diff
@@ -1155,4 +1215,16 @@ private void release() {
...
+
+    @Override
+    public void onMetadataUpdate(Cluster cluster) {
+        final List<String> topicsToSubscribe = new ArrayList<>();
+
+        for (String topic : cluster.topics())
+            if (this.subscriptions.getSubscribedPattern().matcher(topic).matches())
+                topicsToSubscribe.add(topic);
+
+        subscribeTopics(topicsToSubscribe, this.subscriptions.listener().underlying(), true);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Passing in the current listener seems a little weird. I wonder if we're trying a little too hard to reuse the subscribeTopics method in SubscriptionState. If instead we had a method like SubscriptionState.subscribeMatchingTopic (or something like that), then we wouldn't need the flag and we wouldn't need to pass a listener. You could change this line to this:

```
subscription.subscribeMatchingTopics(topicsToSubscribe);
metadata.setTopics(topicsToSubscribe)
```

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 12]
```diff
@@ -644,10 +646,14 @@ private KafkaConsumer(ConsumerConfig config,
      */
     @Override
     public void subscribe(List<String> topics, ConsumerRebalanceListener listener) {
+        subscribeTopics(topics, listener, false);
+    }
+
+    private void subscribeTopics(List<String> topics, ConsumerRebalanceListener listener, boolean isSubscribedViaPatternSubscription) {
         acquire();
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: It's a minor thing, but I would move the acquire/release to the public subscribe method.

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java (hunk 11 above)
Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java]: Makes sense.

**Comment by SinghAsDev** [comment]
Unrelated test failure. @hachikuji could you take a look. Thanks!

**Comment by hachikuji** [comment]
@SinghAsDev Looks good for the most part. We just need to figure out what to do with unsubscribe. We could have `unsubscribe(Pattern)` or `unsubscribe()`, or no unsubscribe at all. If the latter, then we could  have either subscribe(topic) or subscribe(pattern) always override the previous subscription, which would make `subscribe(Collections.emptyList)` always have the same effect as `unsubscribe()`. @guozhangwang Do you have any thoughts?

**Comment by guozhangwang** [comment]
LGTM.

About unsubscribing, I think it is useful to keep the `unsubscribe()` to unsubscribe any topic / partition / pattern instead of keeping a `unsubscribe(Pattern)` since we will only have one pattern subscribed at the same time.

**Comment by SinghAsDev** [comment]
@hachikuji @guozhangwang changed the behavior of unsubscribe. Up for your reviews again :)

File: clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java [hunk 13]
```diff
@@ -674,6 +676,50 @@ public void subscribe(List<String> topics) {
...
+    }
+
+    /**
+     * Unsubscribe from topics currently subscribed to
+     */
+    public void unsubscribe() {
+        acquire();
+        try {
+            this.subscriptions.unsubscribe();
+            this.metadata.needMetadataForAllTopics(false);
```
Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:715]: To be clear, we'll continue fetching metadata for all the topics that were added to Metadata directly. Is that right? If instead we cleared those topics, would the NetworkClient fetch metadata for all topics until we've set a new subscription?

Comment (guozhangwang) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:715]: I think this is a bug in the protocol that if the metadata request contains empty list, then returns the full topic list. This should be fixed in another ticket.

Another two general comments: 1) we could remove a duplicate ConsumerRebalanceListener / RebalanceListener, and then remove the Consumer parameter in the callback functions; 2) we probably can create a member Metadata.Listener inside KafkaConsumer instead of letting KafkaConsumer to implement this interface along with Consumer, since it is only used in subscribe(Pattern).

Just to not drag this ticket further I suggest we do these in later tickets. @hachikuji Does that sound good to you?

Comment (hachikuji) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:715]: @guozhangwang Yep, sounds good to me.

Comment (SinghAsDev) [inline clients/src/main/java/org/apache/kafka/clients/consumer/KafkaConsumer.java:715]: @guozhangwang @hachikuji thanks for the reviews and helping in getting this committed. If there is no JIRA already I can create the JIRA and PR for fixing the issues mentioned by @guozhangwang . Let me know. Thanks!

**Comment by hachikuji** [comment]
@SinghAsDev I just created a ticket for removing the consumer instance from the callback. You can create a ticket for the other issue if you want.

**Comment by SinghAsDev** [comment]
@hachikuji I have created KAFKA-2533, will create PR shortly.
